### PR TITLE
Testing harness no events for own data

### PR DESCRIPTION
**Pull request by jameinel**
This is an adaptation of pull #212 .

When using `Harness.update_relation_data` we ensure that you don't trigger an event when updating your own relation data, but we also don't prevent you from doing so in the harness.
The additional parameters to `add_relation`, and `add_relation_unit` are removed, standardizing on using `update_relation_data` to set the data. Data buckets are all initialized to empty.

### CODE REVIEW

File: ops/testing.py [hunk 1]
```diff
@@ -242,25 +233,36 @@ def update_relation_data(self, relation_id, app_or_unit, key_values):
...
-                entity = self._model.get_unit(app_or_unit)
-            else:
-                entity = self._model.get_app(app_or_unit)
-            rel_data = relation.data.get(entity, None)
-            if rel_data is not None:
-                # If we have read and cached this data, make sure we invalidate it
-                rel_data._invalidate()
-        # TODO: we only need to trigger relation_changed if it is a remote app or unit
+
+        if entity == self._model.unit:
```
Comment (chipaca) [inline ops/testing.py:256]: the objects you're comparing here have no comparison methods, so this is the same as saying `entity is self._model.unit`.
It's not clear to me that the fact that the object returned by e.g. `self._model.get_unit` is identical (as in, satisfies `is`) to the `self._model.unit` is guaranteed to be true in the future, or even if it is true by design and isn't just happenstance.

File: ops/testing.py [hunk 2]
```diff
@@ -99,6 +102,7 @@ class TestCharm(self._charm_cls):
         # rather than TestCharm has no attribute foo.
         TestCharm.__name__ = self._charm_cls.__name__
         self._charm = TestCharm(self._framework, self._framework.meta.name)
+        self._begun = True
```
[the remark this answers is not in the document]
Comment (jameinel) [reply ops/testing.py:105]: I missed this part, I think we can get rid of _begun entirely, as the effect of begin is to create _charm. Do you think we need the RuntimeError check, could we just put "if self._charm is not None" ?

File: ops/testing.py (hunk 1 above)
Comment (jameinel) [reply ops/testing.py:256]: I'm happy enough with "if app_or_unit == self._model.unit.name" as we won't have 2 objects with the same name that aren't supposed to be the same.
(done)

File: ops/testing.py [hunk 3]
```diff
@@ -242,25 +231,36 @@ def update_relation_data(self, relation_id, app_or_unit, key_values):
...
         """
+        relation_name = self._backend._relation_names[relation_id]
+        relation = self._model.get_relation(relation_name, relation_id)
+        if '/' in app_or_unit:
+            entity = self._model.get_unit(app_or_unit)
+        else:
+            entity = self._model.get_app(app_or_unit)
+        rel_data = relation.data.get(entity, None)
+        if rel_data is not None:
+            # If we have read and cached this data, make sure we invalidate it
```
Comment (niemeyer) [inline ops/testing.py:242]: Isn't the _get_ above going to read and cache it?

Comment (jameinel) [reply ops/testing.py:242]: _load (which calls 'relation-get') is triggered when you access an attribute (relation.data[foo]._data). 

So accessing rel_data[key] actually triggers the load, not grabbing the rel_data object.

Comment (jameinel) [reply ops/testing.py:242]: I probably didn't state that as well as I could.
`relation.data.get(entity)`
just returns the RelationDataContent object. That object is a LazyMapping so it doesn't actually call _load (triggering 'relation-get') until you access an attribute (either len(), iter() or __getitem__).

So it doesn't actually read relation-get just grabbing the idea-of-an-entity's data.

**Pull request merged**